Plotting - Automatic line coloring when plotting multiple expressions

Davide-sd opened the pull request:
#### Brief description of what is fixed or changed

1. Modified the Line series classes: now they return data which can
be used directly in matplotlib's plt.plot() method, which support
automatic line coloring. Previously, the format of the data was specific to
Matplotlib's `LineCollection`, which unfortunately doesn't support automatic
line coloring. The new format of the data makes it possible to easily create new
backends.
Back-compatibility has been maintained: it is still possible to apply a function to
the `line_color` attribute in order to create a color gradient.

2. Modified the `plot3d_parametric_line`, `plot3d`, `plot3d_parametric_surface`
methods in order to correctly set the axis labels in 3D plots.

3. Added the `zlabel` attribute to the `Plot` class, which will be rendered
by the backend. Similarly to a 2D plot where, for example, the x-label is set to
`"x"` and the y-label is set by default to `"f(x)"`, this edit will create a default
z-label `'"f(x, y)"` in 3D plots.

4. Modified the MatplotlibBackend class to:
	a. allow for automatic line coloring, while keeping back-compatibility
	with the `line_color` attribute.
	b. set the correct axis labels for 3D plots.

5. Updated the doc-strings associated to the `line_color` attribute.

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs

Fixes #19384

<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* plotting
  * Automatic line coloring

<!-- END RELEASE NOTES -->

Davide-sd committed:
Plotting - Automatic line coloring when plotting multiple expressions

1. Modified the Line series classes: now they return data which can
be used directly in matplotlib's plt.plot() method, which support
automatic line coloring. Back-compatibility has been maintained.

2. Modified the plot3d_parametric_line, plot3d, plot3d_parametric_surface
methods, in order to correctly set the axis labels.

3. Added the `zlabel` attribute to the Plot class, which will be rendered
by the backend.

4. Modified the MatplotlibBackend class to:
	a. allow for automatic line coloring, while maintaing back-compatibility
	with the `line_color` attribute of the different plot methods.
	b. set the correct axis labels in 3D plots.

5. Updated the docstrings associated to the `line_color` attribute.

Davide-sd committed:
Removed unused import + fix axis labels

1. Removed `import warnings` in order to pass flake8 tests.
2. Corrected axis labels for plot3d_parametric_line and plot3d_parametric_surface.
By default, the axis labels will be set to `"x"`, "y", f(x, y)`. If the users is
going to set `xlabel="a"` and/or `ylabel="b"`, then zlabel will be `"f(a, b)"`.

Davide-sd committed:
Fixed bug with surface_color

Even though the documentation clearly states that `surface_color` should be a function,
if a user erroneously typed a float number, an error would have been raised, because
the `nb_of_points` attribute is not defined for Surface series classes.
With this change, the error has been solved by using the two attributes defined in
those classes.

sylee957 commented on sympy/plotting/plot.py:
```diff
@@ -2289,6 +2358,9 @@ def plot3d_parametric_surface(*args, show=True, **kwargs):
     series = [ParametricSurfaceSeries(*arg, **kwargs) for arg in plot_expr]
+    xlabel = kwargs.setdefault("xlabel", "x")
+    ylabel = kwargs.setdefault("ylabel", "y")
+    kwargs.setdefault("zlabel", "f(%s, %s)" % (xlabel, ylabel))
```
For parametric line and surface, I think that it should better stay as `z`

Davide-sd replied:
I agree. Done it.

Davide-sd committed:
Setting `zlabel = "z"` for 3d parametric line and surface

sylee957 commented:
I'm not sure why `legacy` keyword should be needed. Is it difficult to convert between list of segments vs list of points  by other means?

Davide-sd commented:
Returning a list of segments is only required to use Matplotlib's `LineCollection`, which allows to use gradient coloring on the lines. So, the current implementation of `LineOver1DRangeSeries` and `Parametric2DLineSeries` is very Matplotlib specific. Note that by using `LineCollection` we get problems when we would like to show the legend, but that's another issue...

As far as I know, most of the other plotting libraries require list of points. Keep in mind that the user can implement their backends, PR #20463 ) (As a matter of fact, I usually use 3 different backeds depending on the problem at hand). Therefore, by using the `legacy` option, we make  `LineOver1DRangeSeries` and `Parametric2DLineSeries` almost backend-independent, hence reducing the code required to implement other backends (which is a very good thing, given Matplotlib limitations).

In my opinion, the series classes should only return list of points, and then `MatplotlibBackend` should implement its conversion function. Also, if it were for me, the `get_color_array` function of the `Line2DBaseSeries` class should also be moved into `MatplotlibBackend`. Unfortunately, that would break back-compatibility, that's why I created the `legacy` option.

oscarbenjamin commented:
It would be better if sympy provided a function for getting the list of points in the first place and then the user could send them to whatever plotting library they like. Then sympy could provide high-level convenience routines for making some plots but there wouldn't be any need for a backend architecture because anyone who wanted something more complicated could easily do it themselves.

Davide-sd commented:
Nope. The beauty of the current plotting module is the capability to create simple plots "independently" of the backend with just a single plot command. 

Sympy definitely needs new functions to easily convert a symbolic expression into numerical data that can be used by any plotting library, and part of that is already implemented in the Series classes. If the user find that any of the simple plot commands do not suit its needs, then it can use these functions and insert the numerical data into whatever library he wants.

oscarbenjamin commented:
> with just a single plot command.

To be clear I don't object to having a single command like `plot(expr, use='plotly')`. What I object to is having that function return a wrapper class that tries to wrap the other libraries internally while hiding their native API.

sylee957 commented:
I think that it's better to update `get_points`  instead of modifying get_segments if you want to have a function to get points. 
`legacy` seems like an awkward keyword that indicates some deprecation need to be done, but doesn't follow https://github.com/sympy/sympy/wiki/Deprecating-policy

Davide-sd commented:
> I think that it's better to update get_points instead of modifying get_segments if you want to have a function to get points.

That's easier said than done, because the actual `get_points` is used when `adaptive=False`. To do what you want it to do, it would require a substantial change in the current implementation of the Line Series classes, which would definitely break backcompatibility.

I think we can try the following:
* Move the adaptive algorithm implemented in `LineOver1DRangeSeries.get_segments` and `Parametric2DLineSeries.get_segments` into a new function `get_data`, which returns list of points.
* Remove the `legacy` option. Then, `get_segments` would first call `get_data` and convert the list of points to a list of segments. In doing so, backcompatibility should be maintained. Since this is a Matplotlib-specific feature, I would also raise a deprecation warning inside this function, and create an equivalent function in `MatplotlibBackend`.

In a later PR, the `get_data` function could also be implemented in the remaining Series classes. In doing so we can easily extract numerical data. For example, it should be possible to do:

```
# get numerical data to create custom plots starting from symbolic expressions
def plot_data(*args, **kwargs):
    args = list(map(sympify, args))
    # automatically create the series by looking at the form of the arguments and the number of free symbols
    # (this is possible, I've already tested it)
    s = _build_series(args, **kwargs)
    # simply call the common interface
    return s.get_data()
```

Without `get_data`, we would have to do something like:
```
# get numerical data to create custom plots starting from symbolic expressions
def plot_data(*args, **kwargs):
    args = list(map(sympify, args))
    # automatically create the series by looking at the form of the arguments and the number of free symbols
    # (this is possible, I've already tested it)
    s = _build_series(args, **kwargs)
    # simply call the common interface
    if s.is_implicit:
	return s.get_raster()
    elif s.is_3Dsurface:
	return s.get_meshes()
    ...
```

Davide-sd committed:
Removed legacy option

This commit removes the previously introduced `legacy` option in the `get_segments` methods
and introduced the `get_data` method.

`get_data` returns list of coordinates which can directly be used by most plotting libraries.
`get_segments` convert the list of coordinates generated by `get_data` into a list of segments,
which is used by Matplotlib in order to apply color-gradients to the lines.

This commit doesn't break backcompatibility.

Davide-sd committed:
Fixed unindent in order to pass Sphinx tests

Davide-sd committed:
Removed unnecessary method

Davide-sd committed:
Added first point to the parametric adaptive algorithm

In the previous commit I modified the code to return lists of points instead of list of segments,
but I forgot to insert the first point to the resulting lists. Hence, when the parametric_plot
would result in a missing point at the start of the curve. This commit fixes it.

sylee957 commented on sympy/plotting/plot.py:
```diff
@@ -947,7 +1005,7 @@ def get_color_array(self):
                 return f(*variables)
         else:
-            return c*np.ones(self.nb_of_points)
+            return c*np.ones(min(self.nb_of_points_x, self.nb_of_points_y))
```
Does this work with ParametricSurfaceSeries? I see it uses `nb_of_points_u, nb_of_points_v` instead of `x, y`

Davide-sd replied:
Nope, need to fix it!

sylee957 replied:
I think that you should add some dry-run tests (without checking image output) in test_plot to see if these don’t raise some exceptions at least

Davide-sd replied:
Done. Thank you for pointing that out! :)

Davide-sd committed:
Updated MatplotlibBackend to implement its own get_segments

The function get_segments currently implemented in the Line2DBaseSeries returns data specifically
arranged to be used in MatplotlibBacked. Hence, it makes sense that this function is actually
implemented in that class.

Davide-sd committed:
Fixed trailing white spaces

Davide-sd committed:
Fixed gradient coloring for 3D parametric line

Davide-sd committed:
Added tests for custom coloring

This commit adds a few test to assert that no errors are raised when custom
coloring is used, that is, the attributes `line_color` or `surface_color`

Davide-sd committed:
Removed unused variables

Davide-sd committed:
Added a deprecation warning for the Line2DBaseSeries.get_segments method

Davide-sd committed:
Updated deprecated version

Davide-sd committed:
Code Refactoring to improve readability

Previously, `get_points` methods was called when `adaptive=False`, hence it was used to
uniformly sample the line series (it returned lists of coordinates).

Now, `get_points` is always called. It will use a uniform or adaptive sampling depending
on the value of the `adaptive` attribute. As before, it will return list of coordinates,
hence back-compatibility is maintained.

Davide-sd committed:
Fixed is_parametric attribute of Parametric3DLineSeries
